### PR TITLE
Add sniffers for HDF5 and MZ5

### PR DESCRIPTION
**Pull request by chambm**
MZ5 is a mass spectrometry data format based on HDF5. HDF5 is easy to sniff since it has a magic number. But sniffing a particular HDF5 format (e.g. astronomy vs. mass spectrometry) requires looking further in the file. With MZ5, a sufficiently distinguishing feature is the combination of the FileInformation group and the deltaMZ key in that group. To access those features, the HDF5 library is required (easily available via h5py). But it's not pure python, so it can be tricky to install. I had no luck getting it to work in galaxy (h5py requires numpy>=1.6.1, while the galaxy numpy egg is 1.6.0!), so I implemented a workaround using file extension. Since the sniffers didn't originally have access to the original file name, I added that capability (the most controversial part of this PR, I'm sure). Seems like a reasonable heuristic to me until h5py is available.

### CODE REVIEW

**Comment by bgruening** [comment]
@chambm can we simply assume the h5py library is installed and check (try/except) for this library, with a fallback to your file extension hack?

**Comment by nsoranzo** [comment]
Hi @chambm,
thanks a lot for the pull request! I like the H5 sniffer, but I'm not sure the "original file name" heuristic for MZ5 is a good choice. It is also breaking the unit tests: https://travis-ci.org/galaxyproject/galaxy/builds/75510325

**Comment by chambm** [comment]
@nsoranzo I have a fix for the CI breakage (I think, I can't figure out how to test it on my local machine, even when I follow the commands in the CI log). But it's simply passing an empty string for 'original_filename' from the other places that sniff() is called in sniff.py. I couldn't see any way from that file to get the original_filename, so if sniffing happens in there it can't use the heuristic. Can you elaborate on why the heuristic is not a good choice as a workaround for h5py not being available? It's not hard to imagine another format where there is no feasible way to sniff it without a 3rd party library that doesn't have python bindings.

**Comment by dannon** [comment]
@chambm Regarding running things locally, you should just be able to run [tox](https://tox.readthedocs.org/en/latest/) from the root of your galaxy directory.

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -794,7 +827,7 @@ def set_meta( self, dataset, overwrite=True, **kwd ):
         except Exception as e:
             log.warn( '%s, set_meta Exception: %s', self, e )
 
-    def sniff( self, filename ):
+    def sniff( self, filename, original_name ):
         if super( GeminiSQLite, self ).sniff( filename ):
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/binary.py:789]: Here you should also pass `original_name` to the superclass `sniff()` method.

**Comment by chambm** [comment]
When I run tox, I get:

> galaxy@vagrant-ubuntu-trusty-64:/export/src/galaxy$ tox
> py27-lint installed: flake8==2.4.1,mccabe==0.3.1,pep8==1.5.7,pyflakes==0.8.1,wheel==0.24.0
> py27-lint runtests: PYTHONHASHSEED='29539650'
> py27-lint runtests: commands[0] | bash .ci/flake8_wrapper.sh
> ./test/functional/tools/upload.py:1:2: E901 SyntaxError: invalid syntax
> ./test/functional/tools/upload.py:1:3: E226 missing whitespace around arithmetic operator
> ./test/functional/tools/upload.py:1:6: E226 missing whitespace around arithmetic operator
> ./test/functional/tools/upload.py:1:9: E226 missing whitespace around arithmetic operator
> ./test/functional/tools/upload.py:1:15: E226 missing whitespace around arithmetic operator
> ./test/functional/tools/upload.py:1:27: E226 missing whitespace around arithmetic operator
> ./test/functional/tools/upload.py:1:37: W292 no newline at end of file
> ERROR: InvocationError: '/bin/bash .ci/flake8_wrapper.sh'
> py26-lint create: /home/vagrant/galaxy-tests/py26-lint
> ERROR: InterpreterNotFound: python2.6
> py27-unit installed: wheel==0.24.0
> py27-unit runtests: PYTHONHASHSEED='29539650'
> py27-unit runtests: commands[0] | bash run_tests.sh -u
> Doctest: galaxy.model.User.user_template_environment ... ok
> Doctest: galaxy.objectstore.DiskObjectStore ... ok
> Doctest: galaxy.util.Params ... ok
> Doctest: galaxy.util.file_iter ... ok
> Doctest: galaxy.util.is_binary ... ok
> Doctest: galaxy.util.is_uuid ... ok
> Doctest: galaxy.util.mask_password_from_url ... ok
> Doctest: galaxy.util.merge_sorted_iterables ... ok
> Doctest: galaxy.util.nice_size ... ok
> Doctest: galaxy.util.ready_name_for_url ... ok
> Doctest: galaxy.util.unique_id ... ok
> 
> ---
> 
> Ran 11 tests in 14.863s
> 
> OK
> py26-unit create: /home/vagrant/galaxy-tests/py26-unit
> ERROR: InterpreterNotFound: python2.6
> qunit installed: wheel==0.24.0
> qunit runtests: PYTHONHASHSEED='29539650'
> qunit runtests: commands[0] | bash run_tests.sh -q
> Grunt not on path, cannot run these tests.
> ERROR: InvocationError: '/bin/bash run_tests.sh -q'
> ____________________________________________________________________ summary _____________________________________________________________________
> ERROR:   py27-lint: commands failed
> ERROR:   py26-lint: InterpreterNotFound: python2.6
>   py27-unit: commands succeeded
> ERROR:   py26-unit: InterpreterNotFound: python2.6
> ERROR:   qunit: commands failed

Only 11 tests, what's up with that?

**Comment by nsoranzo** [comment]
Commit c8713582c9797d162cb7be25385febd71b5d18d9 should not be in this branch/PR.

**Comment by chambm** [comment]
I know. Learning about this git thing as I go along. :8ball:

It occurs to me that I could set my tool to accept files with .mz5 extension without doing a proper check to see if they're really mz5 files, just that they're HDF5. Is that a better route until h5py is available?

**Comment by martenson** [comment]
@chambm do you consider this ready for review? 
(also please fix the minor issue in travis)

**Comment by chambm** [comment]
@martenson I think we need some discussion about the file extension heuristic. I also need help getting tox tests to run locally (as I detailed above). And I'd like to get h5py added to the default distro, how can I do that? If h5py was added, I'd forego the file extension heuristic (even though it might be useful in other cases).

**Comment by martenson** [comment]
@chambm If you rebase or merge dev into this our Jenkins server will run the -api and -framework tests for you. If you want to add datatype to distro I recommend creating a separate atomic PR with it against dev.

**Comment by chambm** [comment]
@martenson Are you saying I shouldn't expect to be able to run the tox tests locally? Anyway, I'm thinking I need to split out idpDB, mz5 w/ h5py, and mz5 w/ file extension fallback into separate PRs. The first two should be pretty uncontroversial, but the last one is what needs discussion. Is a "rebase" what I want to accomplish that, or should I just start with two fresh branches, one for idpDB and one for mz5+h5py, then copy the changes over manually?

**Comment by martenson** [comment]
@chambm you can run the tox locally but it is easier to let Jenkins, which is configured to run it always in clean dockerized environment, run it for you 
As of the split I recommend creating new branches and transferring the atomic changes in there, either manually or via cherrypick (if you have it commit-specific).
Let me know if you want more hints. And thank you!

**Pull request closed without merging**